### PR TITLE
8271420: Extend CDS custom loader support to Windows platform

### PR DESCRIPTION
**Pull request by yminqi**
Currently CDS does not support custom loaders on Windows, this patch extends  the support to Windows 64 bit platform.

Tests: mach5 tier1-4

Thanks
Yumin

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8271420](https://bugs.openjdk.java.net/browse/JDK-8271420): Extend CDS custom loader support to Windows platform

### Reviewers
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6202/head:pull/6202` \
`$ git checkout pull/6202`

Update a local copy of the PR: \
`$ git checkout pull/6202` \
`$ git pull https://git.openjdk.java.net/jdk pull/6202/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6202`

View PR using the GUI difftool: \
`$ git pr show -t 6202`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6202.diff">https://git.openjdk.java.net/jdk/pull/6202.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/heapShared.cpp [hunk 1]
```diff
@@ -280,10 +280,10 @@ oop HeapShared::archive_object(oop obj) {
     return ao;
   }
 
-  int len = obj->size();
+  int len = (int)obj->size();
```
Comment (iklam) [inline src/hotspot/share/cds/heapShared.cpp:283]: This code is used by `INCLUDE_CDS_JAVA_HEAP` and is unrelated to custom class loaders.

Comment (calvinccheung) [inline src/hotspot/share/cds/heapShared.cpp:283]: `obj->size()` already returns an `int` so I don't think the typecast is needed.
I agree with Ioi that this file shouldn't be changed.

File: test/hotspot/jtreg/runtime/cds/appcds/test-classes/CustomAppLoader.java [hunk 2]
```diff
@@ -40,7 +40,7 @@
     // args[2...] = arguments for the main class
     public static void main(String args[]) throws Throwable {
         File f = new File(args[0]);
-        URL[] classLoaderUrls = new URL[] {new URL("file://" + f.getCanonicalPath())};
+        URL[] classLoaderUrls = new URL[] {f.toURI().toURL()};
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/appcds/test-classes/CustomAppLoader.java:43]: Not sure if the above can handle relative path.
It is better to set f as follows:
`File f = Path.of(args[0]).toRealPath().toFile();`
Before the above, there should be a check if the length of args is 1.

**Review by calvinccheung — CHANGES_REQUESTED** [review]
I have couple of comments.

File: src/hotspot/share/cds/heapShared.cpp (hunk 1 above)
Comment (yminqi) [reply src/hotspot/share/cds/heapShared.cpp:283]: oop.hpp:105
  // Returns the actual oop size of the object
  inline size_t size();

Comment (yminqi) [reply src/hotspot/share/cds/heapShared.cpp:283]: I will revert the change since it is not related to custom loader.

**Review by calvinccheung — APPROVED** [review]

**Review by iklam — APPROVED** [review]
LGTM

**Comment by yminqi** [comment]
@iklam @calvinccheung Thanks for review!

**Pull request closed without merging**